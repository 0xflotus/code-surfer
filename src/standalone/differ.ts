import { diffLines, applyPatch } from "diff";
import tokenize from "./tokenizer";
import { InputStep } from "code-surfer-types";
const newlineRe = /\r\n|\r|\n/;

function myDiff(oldCode: string, newCode: string) {
  const changes = diffLines(oldCode || "", newCode);

  let oldIndex = -1;
  return changes.map(({ value, count, removed, added }) => {
    const lines = value.split(newlineRe);
    // check if last line is empty, if it is, remove it
    const lastLine = lines.pop();
    if (lastLine) {
      lines.push(lastLine);
    }
    const result = {
      oldIndex,
      lines,
      count,
      removed,
      added
    };
    if (!added) {
      oldIndex += count || 0;
    }
    return result;
  });
}

function insert<T>(array: T[], index: number, elements: T[]) {
  return array.splice(index, 0, ...elements);
}

type Line = {
  content: string;
  slides: number[];
  tokens: { type: string; content: string }[];
};

function slideDiff(
  lines: Line[],
  codes: string[],
  slideIndex: number,
  language: string
) {
  const prevLines = lines.filter(l => l.slides.includes(slideIndex - 1));
  const prevCode = codes[slideIndex - 1] || "";
  const currCode = codes[slideIndex];

  const changes = myDiff(prevCode, currCode);

  changes.forEach(change => {
    if (change.added) {
      const prevLine = prevLines[change.oldIndex];
      const addAtIndex = lines.indexOf(prevLine) + 1;
      const addLines = change.lines.map(content => ({
        content,
        slides: [slideIndex],
        tokens: []
      }));
      insert(lines, addAtIndex, addLines);
    } else if (!change.removed) {
      for (let j = 1; j <= (change.count || 0); j++) {
        prevLines[change.oldIndex + j].slides.push(slideIndex);
      }
    }
  });

  const tokenLines = tokenize(currCode, language);
  const currLines = lines.filter(l => l.slides.includes(slideIndex));
  currLines.forEach((line, index) => (line.tokens = tokenLines[index]));
}

export function parseLines(codes: string[], language: string) {
  const lines: Line[] = [];
  for (let slideIndex = 0; slideIndex < codes.length; slideIndex++) {
    slideDiff(lines, codes, slideIndex, language);
  }
  return lines;
}

export function getSlides(codes: string[], language: string) {
  // codes are in reverse cronological order
  const lines = parseLines(codes, language);
  // console.log("lines", lines);
  return codes.map((_, slideIndex) => {
    return lines
      .map((line, lineIndex) => ({
        content: line.content,
        tokens: line.tokens,
        isNew: !line.slides.includes(slideIndex + 1),
        show: line.slides.includes(slideIndex),
        key: lineIndex
      }))
      .filter(line => line.show);
  });
}

export function getCodes(rawSteps: InputStep[]) {
  const codes: string[] = [];

  rawSteps.forEach((s, i) => {
    if (s.lang === "diff" && i > 0) {
      codes[i] = applyPatch(codes[i - 1], s.code);
    } else {
      codes[i] = s.code;
    }
  });

  return codes;
}
