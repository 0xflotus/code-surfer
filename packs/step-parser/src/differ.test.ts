import { linesDiff, generateIds } from "./differ";

describe("generate line ids", () => {
  it("works with empty array", () => {
    const ids: number[] = [];
    const newIds = generateIds(ids, undefined, 1);
    expect(newIds).toEqual([0.5]);
    expect(ids).toEqual([0.5]);
  });
  it("works with empty array and several new", () => {
    const ids: number[] = [];
    const newIds = generateIds(ids, undefined, 3);
    expect(newIds).toEqual([0.25, 0.5, 0.75]);
    expect(ids).toEqual([0.25, 0.5, 0.75]);
  });
  it("works before existing ids", () => {
    const ids = [0.5];
    const newIds = generateIds(ids, undefined, 3);
    expect(newIds).toEqual([0.125, 0.25, 0.375]);
    expect(ids).toEqual([0.125, 0.25, 0.375, 0.5]);
  });
  it("works after existing ids", () => {
    const ids = [0.5];
    const newIds = generateIds(ids, 0.5, 3);
    expect(newIds).toEqual([0.625, 0.75, 0.875]);
    expect(ids).toEqual([0.5, 0.625, 0.75, 0.875]);
  });
  it("works between existing ids", () => {
    const ids = [0.2, 0.6];
    const newIds = generateIds(ids, 0.2, 3);
    expect(newIds).toEqual([0.3, 0.4, 0.5]);
    expect(ids).toEqual([0.2, 0.3, 0.4, 0.5, 0.6]);
  });
});

describe("differ", () => {
  it("works", () => {
    const codes = [
      `
  console.log(1)
  console.log(2)
  console.log(3)
    `.trim(),
      `
  console.log(1)
  console.log(4)
      `.trim(),
      `
  console.log(1)
  console.log(3)
  console.log(4)
        `.trim(),
      `
  console.log(1)
  console.log(3)
  console.log(4)
          `.trim()
    ];
    expect(linesDiff(codes)).toMatchSnapshot();
  });

  it("works with empty old code", () => {
    const codes = [
      ``,
      `
  console.log(1)
  console.log(4)
      `.trim(),
      ``
    ];
    expect(linesDiff(codes)).toMatchSnapshot();
  });

  it("works when last line changes", () => {
    const codes = [
      `
  console.log(1)
      `.trim(),
      `
  console.log(1)
  console.log(4)
      `.trim()
    ];
    expect(linesDiff(codes)).toMatchSnapshot();
  });

  it("works ?", () => {
    const codes = [
      `var x1 = 1

console.log(x1)
debugger

`,
      ""
    ];
    expect(linesDiff(codes)).toMatchSnapshot();
  });
});
