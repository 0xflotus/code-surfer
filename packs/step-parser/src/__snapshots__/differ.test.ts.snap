// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`differ works ? 1`] = `
Object {
  "lineIds": Array [
    0.2,
    0.4,
    0.6,
    0.8,
  ],
  "steps": Array [
    Array [
      0.2,
      0.4,
      0.6,
      0.8,
    ],
    Array [],
  ],
}
`;

exports[`differ works 1`] = `
Object {
  "lineIds": Array [
    0.25,
    0.3125,
    0.375,
    0.5,
    0.75,
  ],
  "steps": Array [
    Array [
      0.25,
      0.5,
      0.75,
    ],
    Array [
      0.25,
      0.375,
    ],
    Array [
      0.25,
      0.3125,
      0.375,
    ],
    Array [
      0.25,
      0.3125,
      0.375,
    ],
  ],
}
`;

exports[`differ works when last line changes 1`] = `
Object {
  "lineIds": Array [
    0.5,
    0.75,
  ],
  "steps": Array [
    Array [
      0.5,
    ],
    Array [
      0.5,
      0.75,
    ],
  ],
}
`;

exports[`differ works with empty old code 1`] = `
Object {
  "lineIds": Array [
    0.3333333333333333,
    0.6666666666666666,
  ],
  "steps": Array [
    Array [],
    Array [
      0.3333333333333333,
      0.6666666666666666,
    ],
    Array [],
  ],
}
`;
